### PR TITLE
Implemented LRT functionality

### PR DESCRIPTION
**Pull request by menoldmt**
- Updated R library to elegantly handle DESeq2's support for the LRT via the rnaseq.Rmd.
- In rnaseq.Rmd, in chunks `dds_list` and `results_05`, see Example 5 for how to include the LRT in your analysis.
- In test library, added tests for making LRT containing dds objects and all practical combinations of parameters that could be passed to `lcdbwf::make_results`
- Help text containing a summary and link to the relevant DESeq2 vignette section has been added to text.yaml in workflows/rnaseq/downstream.
- Documentation has been added to [rnaseq-rmd.rst](https://github.com/lcdb/lcdb-wf/pull/396/files#diff-8b5525bb8ed0c451b92504702317edff0129dbe69623bd61905b49994f31dcceR263)

### CODE REVIEW

File: lib/lcdbwf/tests/testthat/test-contrasts.R [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+  expect_true(identical(names(res), c('res', 'dds', 'label')))
+  expect_true(any(grepl('Wald', mcols(res$res)$description[4])))
+}) # test_that
+# ---------------------------------------------------------------- #
+
+
+# ---- make_results with dds_lrt but with wald test specified ---- #
+# Similar structure to the ASCII table depicted tests from above
+# but with the 'test' argument included in make_results
+# and test == 'LRT'
```
Comment (daler) [inline lib/lcdbwf/tests/testthat/test-contrasts.R:149]: This comment shows up many times below. I'm not clear on how the structure of the function call is similar to the table. Maybe you could rearrange the arguments to better match the table? Or maybe the function calls are straightforward enough that the comment is not necessary (unlike the `if` clauses above when creating results which I think benefits from pasting the table lines).

Rearranging args to match table:

```r
 expect_error(make_results(test='Wald', type=NULL, dds_name='dds_lrt', label='Shrink lrt results'),
              "The 'test' passed to make_results was set to 'Wald' but 'LRT' has been detected in dds"
```

**Review by daler — CHANGES_REQUESTED** [review]
First round of comments; I'll submit another round when I get a chance to look through the rest.

File: lib/lcdbwf/tests/testthat/test-contrasts.R [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+#|    2    | LRT   | normal| -    | -        |   F   |
+#|    5    | Wald  | NULL  | -    | yes      |   C   |
+#|    3    | Wald  | ashr  | -    | yes      |   A   |
+#|    7    | Wald  | apeglm| yes  | -        |   B   |
+#|    3    | Wald  | normal| -    | yes      |   C   |
+#|    6    | NULL  | NULL  | -    | yes      |   C   |
+#|    4    | NULL  | ashr  | -    | yes      |   A   |
+#|    8    | NULL  | apeglm| yes  | -        |   B   |
+#|    4    | NULL  | normal| -    | yes      |   C   |
+#+---------+-------+-------+------+----------+-------+
```
Comment (daler) [inline lib/lcdbwf/tests/testthat/test-contrasts.R:42]: this table is great!

File: lib/lcdbwf/tests/testthat/test-contrasts.R [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+      res <- make_results(dds_name='dds_wald',
+                          label='Shrink Wald results',
+                          contrast=contrast,
+                          type=type,
+                          test='Wald')
+      expect_true(is_deseq_res(res$res))
+      expect_true(identical(names(res), c('res', 'dds', 'label')))
+      expect_true(!all(res$res$log2FoldChange == 0))
+      expect_true(any(grepl('Wald', mcols(res$res)$description[4])))
+    } else if (type == 'apeglm') {
```
Comment (daler) [inline lib/lcdbwf/tests/testthat/test-contrasts.R:221]:
```suggestion
    # Can't use `contrast` argument for apeglm
    } else if (type == 'apeglm') {
```

File: lib/lcdbwf/tests/testthat/test-contrasts.R [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+      expect_true(!all(res$res$log2FoldChange == 0))
+      expect_true(any(grepl('Wald', mcols(res$res)$description[4])))
+    } # if type
+  }) # test_that
+} # for type
+# ---------------------------------------------------------------- #
+
+
+# ---------------------- missing dds_list ------------------------ #
+remove(dds_list)
```
Comment (daler) [inline lib/lcdbwf/tests/testthat/test-contrasts.R:238]: might want to add it back in later after this test, otherwise adding new tests to the bottom of this file in the future may fail.

File: lib/lcdbwf/tests/testthat/test-dds.R [hunk 5]
```diff
@@ -1,17 +1,119 @@
-test_that("stripping gene versions", {
+#library(DESeq2)
+#library(testthat)
+#devtools::load_all('../../../../lib/lcdbwf')
```
Comment (daler) [inline lib/lcdbwf/tests/testthat/test-dds.R:3]: Can remove these

File: lib/lcdbwf/tests/testthat/test-dds.R [hunk 6]
```diff
@@ -1,17 +1,119 @@
...
+  design_data <- make_design_data()
+  design_data$test <- 'LRT'
+  design_data$reduced_design <- ~1
+  dds <- make_dds(design_data,
+                  config=config,
+                  featureCounts='featurecounts.txt',
+                  parallel=config$parallel$parallel)
+  # Check that the dds object is a DESeqDataSet (not NULL)
+  expect_true(inherits(dds, "DESeqDataSet"))
+  # Check that the WaldStatistic has been generated
```
Comment (daler) [inline lib/lcdbwf/tests/testthat/test-dds.R:62]:
```suggestion
  # Check that the WaldStatistic has not been generated
```

Come to think of it, the `expect_true` and `expect_false` are clear enough that you prob don't need comments for them at all.

File: lib/lcdbwf/tests/testthat/test-contrasts.R [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+  dds <- dds_list[['dds_wald']]
+  # Directly pass the dds object
+  res <- make_results(dds_name=dds,
+                      label='Direct DDS',
+                      type='ashr',
+                      contrast=contrast)
+  expect_true(is_deseq_res(res$res))
+  expect_true(identical(names(res), c('res', 'dds', 'label')))
+  expect_true(any(grepl('Wald', mcols(res$res)$description[4])))
+}) # test_that
```
Comment (daler) [inline lib/lcdbwf/tests/testthat/test-contrasts.R:142]: Minor style comment: I don't think we need `# test_that` at the end of every single test. I think the current formatting/indentation makes it pretty clear, and it's not multiply nested so I feel like there's low confusion risk.

File: lib/lcdbwf/tests/testthat/test-results.R [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+#library(DESeq2)
+#library(lcdbwf)
+#library(rlang)
+#library(stringr)
+#library(BiocParallel)
+#library(ggplot2)
+#library(AnnotationHub)
+#library(dplyr)
+#register(MulticoreParam(config$parallel$cores))
+#devtools::load_all('../../../../lib/lcdbwf')
```
Comment (daler) [inline lib/lcdbwf/tests/testthat/test-results.R:11]: Can clean up here too

File: lib/lcdbwf/tests/testthat/test-results.R [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+dds_and_res <- make_deseq_results(test='LRT', type=NULL, reduced_design=~1)
+dds <- dds_and_res$dds
+res <- dds_and_res$res$res
+
+lrt_dds_list <- list(dds1=dds)
+lrt_res_list <- list(res1=list(res=res, dds='dds1', label='LRT'))
+lrt_res_list <- lcdbwf:::attach_extra(lrt_res_list, config)
+
+# ------ Test build_results_tabs function ------ #
+test_that("build_results_tabs works with Wald test", {
```
Comment (daler) [inline lib/lcdbwf/tests/testthat/test-results.R:64]: Were there previous issues where this wasn't working? Or maybe LRT wasn't working, and that needed a test, and it makes sense to have a symmetrical test for Wald?

File: lib/lcdbwf/tests/testthat/test-results.R [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    suppressWarnings(build_results_tabs(lrt_res_list, lrt_dds_list, config, text))
+  )
+
+  expect_true(any(grepl("LRT log2FoldChange values have been set to 0", mdcat_output)))
+})
+
+# Test that mdcat is not called with LRT expected values for Wald test
+test_that("build_results_tabs does not call mdcat with LRT expected character for Wald", {
+  # build_results_tabs requires 'dds_list' in .GlobalEnv
+  dds_list <<- wald_dds_list
```
Comment (daler) [inline lib/lcdbwf/tests/testthat/test-results.R:128]: should this be reset at the end of this test so that future tests appended to the bottom don't get `wald_dds_list`? Same for where it's set in the test above (but lrt_dds_list in that case).

File: workflows/rnaseq/downstream/text.yaml [hunk 11]
```diff
@@ -118,12 +118,18 @@ results_plots:
     normalized to library size, not transcript length. We can say that a gene
     with higher baseMean than another gene has more observed reads, but we
     cannot say whether that is because it has a longer transcript or because
-    there are more transcripts.
+    there are more transcripts. For likelihood ratio test results, we impute
```
Comment (daler) [inline workflows/rnaseq/downstream/text.yaml:121]:
```suggestion
    there are more transcripts. For likelihood ratio test results, we set
```

File: workflows/rnaseq/downstream/text.yaml [hunk 12]
```diff
@@ -118,12 +118,18 @@ results_plots:
...
+    the log2FoldChange values to 0. See "log2FoldChange" below and [DESeq2 likelihood ratio test]
+    (https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#i-ran-a-likelihood-ratio-test-but-results-only-gives-me-one-comparison.)
+    for details as to why.
 
   volcano: |
     A **volcano plot** shows fewer dimensions than an M-A plot (it does not show
     normalized counts) but it can be useful for looking at the relationship of
-    adjusted p-value and log2FoldChange.
+    adjusted p-value and log2FoldChange. For likelihood ratio test results, we impute
```
Comment (daler) [inline workflows/rnaseq/downstream/text.yaml:129]:
```suggestion
    adjusted p-value and log2FoldChange. For likelihood ratio test results, we set
```

File: workflows/rnaseq/downstream/text.yaml [hunk 13]
```diff
@@ -224,7 +230,16 @@ results_files:
     [apeglm](https://doi.org/10.1093/bioinformatics/bty895),
     [ashr](https://doi.org/10.1093/biostatistics/kxw041), and
     [normal](https://doi.org/10.1186/s13059-014-0550-8). **This is designed to
-    be the most important column to sort on.**.
+    be the most important column to sort on.**. We do not report meaningful
+    LFC values for results generated by the likelihood ratio test (LRT). The
+    LRT is a method for comparing the goodness of fit between a full model and
+    a reduced model, evaluating whether the reduced model adequately explains the
+    data without relying on a single pairwise comparison. The log2FoldChange is
+    only applicable for pairwise comparisons; therefore, we impute the
```
Comment (daler) [inline workflows/rnaseq/downstream/text.yaml:238]:
```suggestion
    only applicable for pairwise comparisons; therefore, we set the
```

**Review by daler — CHANGES_REQUESTED** [review]

**Comment by daler** [comment]
I know it looks like a lot of comments above, but this was a lot of work you did and there's a lot going on here!

This will be a great addition, and the detailed tests will give us a lot of confidence that it's correct and all the corner cases are handled. Nice work.

On the CI side, the `test_that` tests are running, but if they fail we get a false pass from CircleCI. I'll figure that out in a separate PR and merge it into this one. Then we'll get the tests to properly pass here.

**Comment by daler** [comment]
Actually I just realized that `rnaseq.Rmd` is not touched in this PR. I think it would be good to include another example in `rnaseq.Rmd`, showing how to run an LRT so that users will know that's even an option. We should also find the right place in the existing documentation to describe these changes.

File: lib/lcdbwf/tests/testthat/test-contrasts.R [hunk 14]
```diff
@@ -0,0 +1,246 @@
+#library(DESeq2)
+#library(testthat)
+#library(rlang)
+#library(stringr)
+#library(BiocParallel)
+#devtools::load_all('../../../../lib/lcdbwf')
+#register(MulticoreParam(config$parallel$cores))
```
[the remark this answers is not in the document]
Comment (daler) [reply lib/lcdbwf/tests/testthat/test-contrasts.R]: Done in dd208bb94fae8f3337

File: lib/lcdbwf/tests/testthat/test-contrasts.R (hunk 4 above)
Comment (daler) [reply lib/lcdbwf/tests/testthat/test-contrasts.R]: addressed in 9828f4690e7bdee5

File: lib/lcdbwf/tests/testthat/test-results.R (hunk 8 above)
Comment (daler) [reply lib/lcdbwf/tests/testthat/test-results.R]: addressed in dd208bb94fae8f333

File: lib/lcdbwf/tests/testthat/test-results.R (hunk 9 above)
Comment (menoldmt) [reply lib/lcdbwf/tests/testthat/test-results.R:53]: I made a separate test for Wald for completions sake but there is no reason build_results_tabs should fail if the results object in res_list was created without error for LRT or Wald

File: lib/lcdbwf/R/results.R [hunk 15]
```diff
@@ -92,7 +92,5 @@ build_results_tabs <- function(res_list, dds_list, config, text){
 #' @return Boolean TRUE if results object's pvalues were provided by
 #' nBinomLRT and FALSE if the Wald test was used.
 check_LRT <- function(res_i) {
-  mcols_pval <- mcols(res_i)$description[9]
-  mcols_pval <- grepl('LRT', mcols_pval)
-  return(mcols_pval)
+  mcols <- grepl('LRT', mcols(res_i)$description[9])
```
Comment (daler) [inline lib/lcdbwf/R/results.R:95]: does this need `S4Vectors::`? Also seems to be overwriting `mcols` (albeit within the function scope, but still could be confusing).

File: lib/lcdbwf/R/contrasts.R [hunk 16]
```diff
@@ -125,6 +125,9 @@ dds_coefs <- function(dds, ..., expand=FALSE){
 #' @param label Label to describe this contrast which will be used in headings.
 #' @param dds_list List of dds objects. If NULL, then look in the global
 #'   environment for an object called "dds_list" and use that.
+#' @param type Type of shrinkage for use by lfcShrink(). If no type is given,
+#'   we use the current DESeq2 default argument for Type. If
```
Comment (daler) [inline lib/lcdbwf/R/contrasts.R:129]:
```suggestion
#'   we use the current DESeq2 default `type` argument for lfcShrink. If
```

File: lib/lcdbwf/R/contrasts.R [hunk 17]
```diff
@@ -154,85 +157,117 @@ make_results <- function(dds_name, label, dds_list=NULL, ...){
...
+  # DDS object contains the LRT, we convert all values in the log2FoldChange
+  # column of the DESeqResults object to 0. LFC values only make sense to report for a single
+  # comparison of two sample groups. This applies to the Wald test only.
+  # LRT is instead performing a test of the removal of one or more factor(s) from the design formula.
+  # DESeq2 reports log2FoldChange values for a single pair-wise comparison when test == 'LRT'. This
+  # can be misleading and so this is our solution.
+
+  # Adjust log2FoldChange for LRT test
+  if (!is.null(dots$test) && dots$test == 'LRT') {
+    res$log2FoldChange <- 0
```
Comment (daler) [inline lib/lcdbwf/R/contrasts.R:209]: Add a warning that gets printed describing what happened, since this may be a surprising side-effect of running this function.

File: lib/lcdbwf/R/contrasts.R [hunk 18]
```diff
@@ -154,85 +157,117 @@ make_results <- function(dds_name, label, dds_list=NULL, ...){
...
+  }
 
   return(
     list(
-      res=res,
-      dds=dds_name,
-      label=label
+      res = res,
+      dds = dds_name,
+      label = label
```
Comment (daler) [inline lib/lcdbwf/R/contrasts.R:236]:
```suggestion
      res=res,
      dds=dds_name,
      label=label
```

File: docs/rnaseq-rmd.rst [hunk 19]
```diff
@@ -260,6 +260,23 @@ adds some extra convenience when working with lists of dds objects, including
 the detection of parallelization as set up in the config object. See the help
 for ``lcdbwf::make_results()`` for more details.
 
+By default, if no test argument is specified in the parameters for
+``lcdbwf::make_dds`` (examples 1-4, rnaseq.Rmd, lines 164-187), the Wald test is
```
Comment (daler) [inline docs/rnaseq-rmd.rst:264]: lines will change over time, so don't include them here.

File: docs/rnaseq-rmd.rst [hunk 20]
```diff
@@ -260,6 +260,23 @@ adds some extra convenience when working with lists of dds objects, including
...
+performed. When ``lcdbwf::make_results`` processes a Wald test dds object, it
+detects the Wald test and expects a ``contrast`` or ``coef`` argument to specify which
+p-values and log2FoldChange values to report.
+
+DESeq2 also supports the nBinomLRT (LRT). Example 5 (rnaseq.Rmd, line 189)
```
Comment (daler) [inline docs/rnaseq-rmd.rst:269]: same here, no line number

**Review by daler — COMMENTED** [review]
looks good, just a couple minor docs fixes and i think this is ready to go

File: lib/lcdbwf/R/contrasts.R [hunk 21]
```diff
@@ -196,17 +196,18 @@ make_results <- function(dds_name, label, dds_list=NULL, ...){
...
-  # comparison of two sample groups. This applies to the Wald test only.
+  # comparison of two sample groups. This only applies to the Wald test.
   # LRT is instead performing a test of the removal of one or more factor(s) from the design formula.
   # DESeq2 reports log2FoldChange values for a single pair-wise comparison when test == 'LRT'. This
   # can be misleading and so this is our solution.
 
   # Adjust log2FoldChange for LRT test
   if (!is.null(dots$test) && dots$test == 'LRT') {
     res$log2FoldChange <- 0
+    warning("All log2FoldChange values in the DESeq2 results object have been set to 0. See https://github.com/lcdb/lcdb-wf/blob/master/docs/rnaseq-rmd.rst?plain=1#L269.")
```
Comment (daler) [inline lib/lcdbwf/R/contrasts.R:210]: Will be better to link to the relevant rendered HTML docs section rather than a line number that will change over time

File: docs/rnaseq-rmd.rst [hunk 22]
```diff
@@ -260,6 +260,23 @@ adds some extra convenience when working with lists of dds objects, including
...
+demonstrates how to create a dds object with LRT data. Since the LRT tests
+the removal of one or more terms from the design formula, a single
+log2FoldChange column doesn't reflect the test's complexity. DESeq2's results
+object is optimized for the Wald test, and when storing LRT results, it
+maintains consistency in datastructure by choosing a single pair-wise comparison for
+log2FoldChange values. To avoid confusion, we set all log2FoldChange values to
+0 for LRT results.
```
Comment (daler) [inline docs/rnaseq-rmd.rst:276]: This is very unexpected behavior, so should be more prominent in the docs

```suggestion
log2FoldChange values. To avoid confusion, ***we set all log2FoldChange values to
0 for LRT results***.
```
